commons/src/main/java/io/aiven/kafka/connect/common/source/AbstractSourceTask.java

```diff
@@ -0,0 +1,510 @@
+        super();
+        this.logger = logger;
+        connectorStopped = new AtomicBoolean();
+        timer = new Timer(MAX_POLL_TIME);
+        backoff = new Backoff(timer.getBackoffConfig());
+    }
+
+    /**
+     * Gets the iterator of SourceRecords. The iterator that SourceRecords are extracted from during a poll event. When
+     * this iterator runs out of records it should attempt to reset and read more records from the backend on the next
+     * {@code hasNext()} call. In this way it should detect when new data has been added to the backend and continue
+     * processing.
+     * <p>
+     * This method should handle any backend exception that can be retried. Any runtime exceptions that are thrown when
+     * this iterator executes may cause the task to abort.
+     * </p>
+     *
+     * @param config
+     *            the configuraiton for the Backoff.
+     * @return The iterator of SourceRecords.
+     */
+    abstract protected Iterator<SourceRecord> getIterator(BackoffConfig config);
+
+    /**
+     * Called by {@link #start} to allows the concrete implementation to configure itself based on properties.
+     *
+     * @param props
+     *            the properties to use for configuration.
+     */
+    abstract protected SourceCommonConfig configure(Map<String, String> props);
+
+    @Override
+    public final void start(final Map<String, String> props) {
+        logger.debug("Starting");
+        config = configure(props);
+        maxPollRecords = config.getMaxPollRecords();
+        sourceRecordIterator = getIterator(timer.getBackoffConfig());
+    }
+
+    /**
+     * Try to add a SourceRecord to the results.
+     *
+     * @param results
+     *            the result to add the recrod to.
+     * @param sourceRecordIterator
+     *            the source record iterator.
+     * @return true if successful, false if the iterator is empty.
+     */
+    private boolean tryAdd(final List<SourceRecord> results, final Iterator<SourceRecord> sourceRecordIterator) {
+        if (sourceRecordIterator.hasNext()) {
+            backoff.reset();
+            final SourceRecord sourceRecord = sourceRecordIterator.next();
+            if (logger.isDebugEnabled()) {
+                logger.debug("tryAdd() : read record " + sourceRecord.sourceOffset());
+            }
+            results.add(sourceRecord);
+            return true;
+        }
+        logger.info("No records found in tryAdd call");
+        return false;
+    }
+
+    /**
+     * Returns {@code true} if the connector is not stopped and the timer has not expired.
+     *
+     * @return {@code true} if the connector is not stopped and the timer has not expired.
+     */
+    protected boolean stillPolling() {
+        final boolean result = !connectorStopped.get() && !timer.isExpired();
+        logger.debug("Still polling: {}", result);
+        return result;
+    }
+
+    @Override
+    public final List<SourceRecord> poll() {
```

Comment:
For clarification: https://kafka.apache.org/37/javadoc/org/apache/kafka/connect/source/SourceTask.html#poll--

This function does comply with the defined API and returns control from time to time to the caller. The question and difference to the defined API is that this does not return `null`, but returns empty list when there is no data available.